### PR TITLE
Add new RandomVariable Op and optimizations

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR adds a more optimizable and robust `Op` for random variables aptly named `RandomVariable`.

- [x] Add `RandomVariable` `Op` and tests
  - Broadcastable multivariate normal, Dirichlet, categorical, and multinomial `Op`s (i.e. they support multiple, stacked independent parameter arguments&mdash;with `size` parameter support, as well)
- [x] Remove `RandomFunction`, its modules, and the functions that depend on it.
- [x] Create `DimShuffle` lift optimization for `RandomVariable`s
  - E.g. `normal(mean, std).T` is replaced with `normal(mean.T, std.T)`
- [x] Create `*Subtensor` lift optimization for `RandomVariable`s
  - E.g. `normal(mean, std)[idx]` is replaced with `normal(mean[idx], std[idx])`

(This is also a replacement for #131 that comes from my new fork)

### CODE REVIEW

**Comment by brandonwillard** [comment]
While attempting to create a lift optimization for `DimShuffle`s on `RandomVariable`s I came across an issue involving numeric reproducibility.  

The problem is neatly summarized by the following NumPy-only example:
```python
>>> np.random.RandomState(123).normal(mean, std).T
array([[0.99989144, 3.99984937],
       [2.00009973, 4.99994214],
       [3.0000283 , 6.00016514]])

>>> np.random.RandomState(123).normal(mean.T, std.T)
array([[0.99989144, 4.00009973],
       [2.0000283 , 4.99984937],
       [2.99994214, 6.00016514]])
```
The first case is the numeric result one would obtain from a `DimShuffle`d `RandomVariable` graph.  The second is the lifted version of the same graph.  Both result are theoretically equivalent and&mdash;ideally&mdash;should produce the same numeric result for the same RNG and seed.  As we can see, they do not.

Here's an example of how it could be made to work:
```python
>>> (mean + std * np.random.RandomState(123).standard_normal((2, 3))).T
array([[0.99989144, 3.99984937],
       [2.00009973, 4.99994214],
       [3.0000283 , 6.00016514]])

>>> mean.T + std.T * np.random.RandomState(123).standard_normal((2, 3)).T
array([[0.99989144, 3.99984937],
       [2.00009973, 4.99994214],
       [3.0000283 , 6.00016514]])
```
Simply put, by implementing the affine transform in `RandomState.normal`, we can add a transpose to the block of standard normals.  This is apparently what we're missing when we use `RandomState.normal`.

Since I don't think we want to effectively reimplement all the samplers in NumPy's `RandomState`, we can either think of a good work around to preserve equality, or we can accept the fact that the two graphs will produce different results although they're theoretically equivalent.

**Comment by brandonwillard** [comment]
A preliminary `Subtensor` lift optimization was just added; however, it needs one addition in order to work with multivariate distributions&mdash;and a lot more tests.

**Comment by brandonwillard** [comment]
OK, it has occurred to me&mdash;in another context&mdash;that we should address the RNG consistency issue mentioned [above](https://github.com/pymc-devs/Theano-PyMC/pull/137#issuecomment-723511546) if we want to apply these optimizations more often.

Imagine that we're creating a Theano graph for the NumPy operations that produce `z` in the following:
```python
import numpy as np

seed = 34893
rng = np.random.RandomState(seed)

x = rng.normal(np.arange(2))

z = x - x[1]
```
```python
>>>  z
array([-0.7960794,  0.       ])
```
Just as with NumPy, we would expect a Theano-PyMC graph for `z` to necessarily have a 0 for the element at index 1.  This should also hold for any RNG state.

The naive `local_subtensor_rv_lift` rewrite rule would effectively substitute `x[1]` with `np.random.RandomState(seed).normal(np.arange(2)[1])`, which would only imply that the _expectation_ of `z[1]` is 0.  I.e.
```python
rng = np.random.RandomState(seed)

x = rng.normal(np.arange(2))

rng_2 = np.random.RandomState(seed)
y = rng_2.normal(np.arange(2)[1])

z = x - y
```
```python
>>>  z
array([-1.       , -0.2039206])
```

Unfortunately, that's not what the graph actually represents, so this rewrite is inconsistent.

As a simple way to avoid introducing this issue, we should not perform the rewrite if there's another reference to `x` in the graph; however, that would limit the applicability of the optimization.  This restriction can be loosened a bit by allowing references to invariant properties (e.g. the shape of `x`) and not the values in `x` themselves.  

We could also attempt to synchronize slices of `x` throughout the graph by replacing the rewritten `RandomVariable`s with stand-ins that are updated in-place; however, this seems fairly complicated.

Alternatively, we might be able to solve a larger number of cases using an RNG-based approach.  Such an approach might also preserve numeric equality (i.e. before and after rewrites), but it will require some additions to the `theano.tensor.random` objects and it will be restricted to only specific RNGs and samplers.  

First, we would need to add support for at least one of the two more flexible NumPy `BitGenerators`: `PCG64` and/or `Philox`.  These RNGs implement an [`.advance`](https://numpy.org/doc/stable/reference/random/bit_generators/generated/numpy.random.PCG64.advance.html#numpy.random.PCG64.advance) method that would allow us to manipulate the state in a manner that preserves consistency between shuffles and subsets of `RandomVariable` arrays.

Our simple example above can be fixed in this way:
```python
x = drng.normal(np.arange(2))

drng = np.random.default_rng(seed)
# Move the state forward so that the next sample matches the second entry in
# `x`
drng.bit_generator.advance(1)
y = drng.normal(np.arange(2)[1])

z = x - y
```
```python
>>>  z
array([-2.68521984,  0.        ])
```

We would need to track the number of elements to skip, but that might not be too difficult in most cases, especially since we're already computing all the requisite shape and index information for the rewrites themselves.

Naturally, this won't work for certain samplers (e.g. rejection-based ones), but it should work for more than a few of the basic random variable samplers.  Otherwise, this approach has a nice correspondence with the actual source of "randomness" (i.e. the RNG), which adds to its appeal.  In other words, lifting the indexing operation (i.e. `rng.normal(np.arange(2))[1]` to `rng.normal(np.arange(2)[1])`) should only affect the `rng` state and nothing else.

**Comment by brandonwillard** [comment]
To keep things moving, we should probably disable the automatic use of these rewrites until a good RNG/rewrite-consistency solution is worked out.  I'll create a separate issue for that.

**Comment by kyleabeauchamp** [comment]
Is there a branch of pymc3 that is compatible with this change?  I'm running into a `ImportError: cannot import name 'MRG_RandomStreams'` exception when I try to use master pymc3 against master theano-pymc.
